### DISCUSSION TITLE
Anti Kick & Steam id Spoof

### DISCUSSION THREAD

**Discussion by D1GQ** [winstxnhdw/lc-hax#399] [General]
I've been messing around with the anti kick for a while, and I've got it to a point where there's a high chance when joining the Lobby you will be a unknown player with a random player steam ID, this isn't what supposed to be happening, I just thought it's pretty interesting there could potentially be a way to spoof your steam ID. And perhaps your name as well.

### Code:
```cs
#pragma warning disable IDE1006

using GameNetcodeStuff;
using HarmonyLib;
using Steamworks;
using System.Reflection;
using Hax;

[HarmonyPatch(typeof(PlayerControllerB), "SendNewPlayerValuesServerRpc")]
class AntiKickPatch {
    static void Prefix(PlayerControllerB __instance) {
        if (!Setting.EnableAntiKick) return;

        // Invoke the original method first
        MethodInfo originalMethod = AccessTools.Method(typeof(PlayerControllerB), "SendNewPlayerValuesServerRpc");
        originalMethod.Invoke(__instance, null);

        // Your modifications go here
        ulong[] playerSteamIds = new ulong[__instance.playersManager.allPlayerScripts.Length];

        for (int i = 0; i < __instance.playersManager.allPlayerScripts.Length; i++) {
            playerSteamIds[i] = __instance.playersManager.allPlayerScripts[i].playerSteamId;
        }

        // Set the Steam ID of the player associated with the host to a spoofed value
        playerSteamIds[__instance.playerClientId] = 0123456789;

        // Invoke the modified method with spoofed Steam IDs
        _ = __instance.Reflect().InvokeInternalMethod("SendNewPlayerValuesClientRpc", playerSteamIds);
    }
}
```

**Comment by winstxnhdw** [comment]
So @armorfid and I tried this. If you spoof your Steam ID, there’s server code that will check for that and then kick you. Were you able to join the game?

**Reply by D1GQ** [reply]
Yes I was

**Reply by winstxnhdw** [reply]
That's cool. I'll look into it further. Thanks!

**Reply by D1GQ** [reply]
Also to note at the time I was using someone's valid steam ID, might be a case where you have to use someone's steam ID to join, I just removed it for privacy reasons.

**Reply by winstxnhdw** [reply]
But you still came in as Nameless?

**Reply by D1GQ** [reply]
Sometimes, but with using that method for some reason I had a higher chance of what @Totoqoe  said to happen, I don't know why though.

**Comment by Totoqoe** [comment]
The random player steam ID is the ID of someone who used to be in the lobby, if you join with antikick and take the player slot that some other player used to have, you'll have the profile of that player.

For example, lets say there's 7 players in a lobby, with slots # 0-6 taken.  Then a player named Amber who had slot # 3 disconnected, if you then join that lobby with antikick, the game will place you in that slot and because you never sent your player values, it recycles the last used ID of the player who had that slot, so now the game thinks I'm "Amber".

**Reply by winstxnhdw** [reply]
That's right! The client is powerless here. It's all in the server.

**Reply by winstxnhdw** [reply]
But I wouldn't say it's completely impossible. We just haven't dug deep enough. We need to know how `Unity.Net` integrates with Steam. I remember I dug really deep and changed the Steam ID at the origin of creation but when I did that, although I managed to spoof my user, the game de-synced when it started and I was no longer able to receive any RPC from the host.

I'd imagine that a real fix to anti-kick, would be a complex multi-step approach, and not anything superficial like simply changing the Steam ID. It would probably require a lot of patching along with changing the Steam ID to achieve it perfectly.

There's many questions to answer.

1. In how many places is the Steam ID set? Is it just at start-up?
2. Why does changing the Steam ID stop me from getting anymore RPCs? Is the RPC socket URI related to the Steam ID? Can we patch that?
3. Is Steam ID all we need to change? When I connected with a spoofed Steam ID, I was labelled "Nameless". Is there a username field we have to populate?

**Reply by Totoqoe** [reply]
I did some testing with a connection emulator, and I noticed that if I limit only my outgoing bandwidth while in the process of joining a lobby, I successfully "join" without sending my join message and I was able to see the players move around, the only thing that went wrong was that I was unable to hear any players and that the game will eventually connect me since I'm not completely blocking the bandwidth. I can usually tell when it does because the audio starts working and the join message appears along with everyones cosmetics. Not sure if this would be some useful info.

**Reply by winstxnhdw** [reply]
Are you sure you are just limiting the bandwidth? How would that work? All requests/RPCs should be atomic and bandwidth would not matter. I won't dismiss this since you've experienced it but what's this tool you are using? It's likely we are missing more context.

**Reply by Totoqoe** [reply]
Yepp, antikick on and bandwidth limited. The tool I was using is called Connection Emulator from Softperfect. I limited my outgoing bandwidth to 8kbps with a queue size of 10000 (meaning the program will not drop packets until it reaches this threshold).
I have not done testing on the other side, I usually have a friend to test some shenanigans with :P

**Reply by winstxnhdw** [reply]
Could you widen the queue size (maybe 10,000,000) and see if this would still occur. It might be dropping packets.

**Reply by Totoqoe** [reply]
The program usually says when it does, I think the issue is that 8kbps is still enough to properly connect you to the lobby, just takes forever to lol.
I can quickly do some testing and video record what it looks like when I do so

**Reply by winstxnhdw** [reply]
That's very interesting because this tells me the game probably has it's own queue size and is dropping the incoming packets. Maybe it's a game network optimisation to do this.

**Reply by Totoqoe** [reply]
Hmmm now my curiosity is peaked, I'll go take a second look lol

**Reply by Totoqoe** [reply]
HOLY SHIT THEY KICKED A WHOLE LOBBY, I don't know if it was the method I did since it seemed like it never broadcasted my join, ill have to check if someone else joined after i did to see if this genuinely excluded me from the list since it would normally lead to a nameless after someone joins.

**Reply by winstxnhdw** [reply]
The host kicked the whole lobby?

**Reply by Totoqoe** [reply]
Watching the footage and it seems someone did, no way the host decided to leave just 1 player after kicking everyone else. Ill post a vid soon.

**Reply by Totoqoe** [reply]
> The host kicked the whole lobby?

yep, then i got the server host error but i know they shut down the lobby because rejoining through the steam ID timed out which means the host closed it after kicking everyone out.

**Reply by Totoqoe** [reply]
https://youtu.be/_Yzfrs3jFJM?feature=shared
Antikick was off btw lol, I did get stuck in the wall but I was still able to receive *and* send data to the host, though I was unable to hear anyone after a certain point (in game audio still worked though)
I set my bandwidth to 3kbps in this video.

**Reply by winstxnhdw** [reply]
`SendConnectionRequest` timeout. Yup, they are dropping packets on their end.

**Reply by Totoqoe** [reply]
So it wasn't anything I did that caused this?

**Reply by winstxnhdw** [reply]
I don’t think so.

**Reply by Totoqoe** [reply]
Awww I thought I found something lol
